### PR TITLE
Navigation: Use getTitle to get the navigation title

### PR DESCRIPTION
**Pull request by scruffian**
## What?
Gets the entity title in a more predictable format. An alternative to https://github.com/WordPress/gutenberg/pull/52167.

## Why?
Sometimes it seems like `useEntityProp` returns an object and sometimes a string, so sometimes this throws an error.

## How?
This changes the approach to use  `getTitle` which is more predictable.

## Testing Instructions
1. Open the site editor
2. Open Library
3. Find a template part that contains a navigation block
4. Open the template part and check that you see the navigation in the left sidebar with the title of the Navigation

### CODE REVIEW

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 /**
  * Internal dependencies
  */
 import SidebarNavigationItem from '../sidebar-navigation-item';
 import { useLink } from '../routes/link';
+import useEditedEntityRecord from '../use-edited-entity-record';
 
 export default function TemplatePartNavigationMenuListItem( { id } ) {
-	const [ title ] = useEntityProp( 'postType', 'wp_navigation', 'title', id );
+	const { getTitle } = useEditedEntityRecord( 'wp_navigation', id );
```
Comment (getdave) [inline packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js]: Seems like a good solution for now.

I wonder if in the future we should have a selector for `getEntityTitle( id )` which does this? It's so common and it causes so many bugs all the time.

I think you should propose it.

Comment (Mamaduka) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js]: This doesn't do anything different internally. See https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-site/src/components/use-edited-entity-record/index.js#L59.

Plus `useEditedEntityRecord` performs extra checks that aren't needed here.

I think we just need to make sure entities return values like titles in a predictable way.

Comment (getdave) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js]: Oh really. Got to be honest I was doing a fly by review. My comment probably wasn't super helpful.

My thought was we need a single source of truth for getting a string-based title. I think the `raw` vs `rendered` thing throws people. Maybe if we had some well-named utilities that got the correct title based on where you want to output it that night help.

`getEntityTitleForField` vs `getEntityTitle` or something. Just an idea...

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js [hunk 2]
```diff
@@ -10,9 +10,9 @@ import TemplatePartNavigationMenuListItem from './template-part-navigation-menu-
 export default function TemplatePartNavigationMenuList( { menus } ) {
 	return (
 		<ItemGroup className="edit-site-sidebar-navigation-screen-template-part-navigation-menu-list">
-			{ menus.map( ( menuId ) => (
+			{ [ ...new Set( menus ) ].map( ( menuId, index ) => (
 				<TemplatePartNavigationMenuListItem
-					key={ menuId }
```
Comment (getdave) [inline packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js:15]: Generally it would be far better to use a unique ID rather than an index as we are already doing. Moreover, [the `Array.map()` method already includes an `index` method on its callback](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#parameters).

Comment (scruffian) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js:15]: Why is a unique id better than an index? Also not that what we are currently doing is not a unique ID as the same navigation can appear on the same template more than once.

Comment (Mamaduka) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js:15]: See: https://react.dev/learn/rendering-lists#why-does-react-need-keys

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js (hunk 1 above)
Comment (scruffian) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js]: I updated this to just use `getEditedEntityRecord`. I'm not sure what the best way to do this is, but right now we have a JS error in trunk and this branch fixes it, so I'd rather get this merged and iterate on improvements later.

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js [hunk 3]
```diff
@@ -10,9 +10,9 @@ import TemplatePartNavigationMenuListItem from './template-part-navigation-menu-
...
+					key={ index }
```
Comment (scruffian) [inline packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js]:
```suggestion
					key={ menuId }
```

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js [hunk 4]
```diff
@@ -10,7 +10,7 @@ import TemplatePartNavigationMenuListItem from './template-part-navigation-menu-
 export default function TemplatePartNavigationMenuList( { menus } ) {
 	return (
 		<ItemGroup className="edit-site-sidebar-navigation-screen-template-part-navigation-menu-list">
-			{ menus.map( ( menuId ) => (
+			{ [ ...new Set( menus ) ].map( ( menuId, index ) => (
```
Comment (scruffian) [inline packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js]:
```suggestion
			{ [ ...new Set( menus ) ].map( ( menuId ) => (
```

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js (hunk 1 above)
Comment (Mamaduka) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js]: I’ll be able to do another round of reviews tomorrow morning.

Can you post more details on how to reproduce issue issue you’re having on trunk?

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js [hunk 5]
```diff
@@ -11,7 +13,20 @@ import SidebarNavigationItem from '../sidebar-navigation-item';
...
+
+		const _record = getEditedEntityRecord(
+			'postType',
+			'wp_navigation',
+			id
+		);
+
+		const templateInfo = getTemplateInfo( _record );
+		return templateInfo?.title || __( '(no title)' );
+	} );
```
Comment (Mamaduka) [inline packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js:29]: Can we just use `record?.title?.rendered || record?.title` here? We can leave `no title` fallback outside of the `mapSelect`.

Using the selector `getTemplateInfo` meant for different entities can have unexpected results in the future.

P.S. The `useSelect` is missing a dependency here.

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js [hunk 6]
```diff
@@ -10,7 +10,7 @@ import TemplatePartNavigationMenuListItem from './template-part-navigation-menu-
 export default function TemplatePartNavigationMenuList( { menus } ) {
 	return (
 		<ItemGroup className="edit-site-sidebar-navigation-screen-template-part-navigation-menu-list">
-			{ menus.map( ( menuId ) => (
+			{ [ ...new Set( menus ) ].map( ( menuId ) => (
```
Comment (Mamaduka) [inline packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js:13]: Is this change related to the title rendering fix?

If we need to ensure that `menus` are a unique set, then this probably has to be done higher in the tree, maybe in `useNavigationMenuContent` or `getBlocksOfTypeFromBlocks`?

Edit: I just saw the #52707. I think this can be removed here.

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js (hunk 1 above)
Comment (scruffian) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list-item.js]: I can't get it to happen since https://github.com/WordPress/gutenberg/pull/52707 merged. Still investigating...

File: packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js (hunk 6 above)
Comment (getdave) [reply packages/edit-site/src/components/sidebar-navigation-screen-pattern/template-part-navigation-menu-list.js:13]: Yes I don't see a new for `Set` here unless I missed something...?

**Comment by scruffian** [comment]
There's a proper fix here: https://github.com/WordPress/gutenberg/pull/52758

**Comment by jeryj** [comment]
The issue this was solving is closed by https://github.com/WordPress/gutenberg/pull/52758
